lib/connection_pool.rb

```diff
@@ -53,22 +53,28 @@ def initialize(options = {}, &block)
   end
 
   def with(options = {})
+    success = false # hoisted
     conn = checkout(options)
     begin
-      yield conn
+      (yield conn).tap do
+        success = true # means the connection wasn't interrupted
+      end
     ensure
-      checkin
+      if success
+        # everything is roses, we can safely check the connection back in
+        checkin
+      else
+        @available.discard!(pop_connection)
```

Comment:
@tamird @mperham Sorry for the dumb question here, but why is the connection considered corrupted at this point?

We use connection pool to share our database connection between test and capybara threads like https://gist.github.com/josevalim/470808#comment-474499

```ruby
class ActiveRecord::Base
  def self.connection
    @@shared_connection || ConnectionPool::Wrapper.new(:size => 1) { retrieve_connection }
  end
end
```

And since this update, we're getting very very weird random errors, mostly in the form of `Mysql2::Error: This connection is in use by: #<Thread:0x007fe86c3263b8 sleep>`.

By adding print statements into the code, I'm seeing we hit this code a lot. What I'm guessing is happening, is the connection is discarded, then recreated when needed again. It uses Rails's internal database connection pool and picks up the same database connection again which is handed off to "other" thread without blocking. Everything goes back to normal again by reverting this ensure block to `checkin`. 

I'm trying to figure out if there's something lacking in this line of code that is causing this problem now. I read through Headius's post and somewhat understand the Timeout::Error problem, but I could use hand pointing out what I'm missing.

Thanks guys.